File: src/mod11_gringotts_concurrency/lesson3_atomic/java/BankAccount.java
package mod11_gringotts_concurrency.lesson3_atomic.java;

import java.util.concurrent.atomic.AtomicInteger;

public class BankAccount {
    private String accountName;
    private AtomicInteger balance;

    public BankAccount(String accountName, int initialBalance) {
        this.accountName = accountName;
        this.balance = new AtomicInteger(initialBalance);
    }

    public BankAccount(String accountName) {
        this.accountName = accountName;
        this.balance = new AtomicInteger(1000);
    }

    public String getAccountName() {
        return accountName;
    }

    public int getBalance() {
        return balance.get();
    }

    public void transferFunds(BankAccount targetAccount, int amount) {
        if (amount <= this.balance.get()) {
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            this.balance.addAndGet(-amount);
            targetAccount.balance.addAndGet(amount);
            System.out.println(
                    amount + " galleons transferred from " + this.accountName + " to " + targetAccount.accountName +
                            "! balance src:" + this.balance + ";balance target:" + targetAccount.getBalance());

        } else {
            System.err.println("Unsufficient funds");
            throw new UnsufficientFundsException();
        }

    }

}

File: src/mod11_gringotts_concurrency/lesson3_atomic/kotlin/BankAccount.kt
package mod11_gringotts_concurrency.lesson3_atomic.kotlin;

import java.util.concurrent.atomic.AtomicInteger

class BankAccount(
    val accountName: String,
    var balance: AtomicInteger = AtomicInteger(1000)

) {
    @Synchronized
    fun transferFunds(targetAccount: BankAccount, amount: Int) {
        if (amount <= this.balance.get()) {
            try {
                Thread.sleep(50)
            } catch (e: InterruptedException) {
                throw RuntimeException(e)
            }
            this.balance.addAndGet(-amount)
            targetAccount.balance.addAndGet(amount)
            println(
                amount.toString() + " galleons transferred from " + this.accountName + " to " + targetAccount.accountName +
                        "! balance src:" + this.balance + ";balance target:" + targetAccount.balance
            )
        } else {
            System.err.println("Unsufficient funds")
            throw UnsufficientFundsException()
        }
    }
}